### PR TITLE
docker_container: wait for removal if removal is in process

### LINKED ISSUE #65811: docker_container with auto_remove:yes fails
##### SUMMARY

`docker_container` module often fails because removal of `auto_remove: yes` takes some time. I had the same problem before with shell scripts since `docker stop` returns immediately but container removal takes some time to finish. Until removal is finished, no new container with the same name can be created.

The error message is:

    Error creating container: 409 Client Error: Conflict ("Conflict. The container name "/MYCONTAINER" is already in use by container "HASH". You have to remove (or rename) that container to be able to reuse that name.")

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
docker_container

##### ANSIBLE VERSION
```
ansible 2.9.2
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/USER/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.17rc1 (default, Oct 10 2019, 10:26:01) [GCC 9.2.1 20191008]

```

##### CONFIGURATION
_no changes_

##### OS / ENVIRONMENT
Ubuntu 19.10

##### STEPS TO REPRODUCE

0. Run playbook to create container
0. Apply many changes to the file system
0. Run playbook again

```yaml
docker_container:
  name: SOMETHING
  image: SOMETHING
  state: started
  detach: yes
  recreate: yes
  autoremove: yes
```

##### EXPECTED RESULTS
Wait for removal before trying to create a new container.

##### ACTUAL RESULTS
Command fails.

```
Error creating container: 409 Client Error: Conflict ("Conflict. The container name "/MYCONTAINER" is already in use by container "HASH". You have to remove (or rename) that container to be able to reuse that name.")
```

### PR DESCRIPTION
**Pull request by felixfontein**
##### SUMMARY
Fixes #65811.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
docker_container

### CODE REVIEW

**Comment by darkdragon-001** [comment]
Tested and working fine for me!

**Comment by felixfontein** [comment]
Great! In that case, it's no longer a WIP ;)

ready_for_review

**Review by Andersson007 — COMMENTED** [review]
@felixfontein , can we reproduce the situation with CI tests?

File: lib/ansible/modules/cloud/docker/docker_container.py [hunk 1]
```diff
@@ -2603,6 +2610,31 @@ def __init__(self, client):
             self.results['ansible_facts'] = {'docker_container': self.facts}
             self.results['container'] = self.facts
 
+    def wait_for_state(self, container_id, complete_states=None, wait_states=None, accept_removal=False):
+        delay = 1.0
```
Comment (Andersson007) [inline lib/ansible/modules/cloud/docker/docker_container.py:2614]:
```
    def wait_for_state(self, container_id, complete_states=None, wait_states=None, accept_removal=False):
        delay = 1.0
```
i wonder if we can implement this as a new module's parameter
(maybe leave this as you implemented in order to backport, then add the option to 2.10)

Comment (felixfontein) [reply lib/ansible/modules/cloud/docker/docker_container.py:2614]: That's what I intend to do :)

**Comment by felixfontein** [comment]
> @felixfontein , can we reproduce the situation with CI tests?

I *knew* you would be asking this ;-) The problem with this is that I didn't even manage to reproduce it locally, so I have no clue on how to reproduce it in CI. Also, I'd guess that reproducing this would induce a high strain on the CI system, potentially making it quite expensive (also in money terms). So I guess it is best to make sure that this doesn't hurt existing use-cases (which CI should do), and hope that @darkdragon-001's successful tests indicate that it actually works in situations where the wait is required.

File: lib/ansible/modules/cloud/docker/docker_container.py (hunk 1 above)
Comment (Andersson007) [reply lib/ansible/modules/cloud/docker/docker_container.py:2614]: Perfect reader’s feedback on a column “You said exactly what i wanted to say but didn’t know how” , Thomas L. Freedman, Thank you for being late

It’s cool, feeling my progress:)

**Comment by Andersson007** [comment]
Ok, clear, if @darkdragoon-001 tested it manually that’s ok (in this case, i said feeling pain a bit:)
I haven’t looked at the code in detail, hope to have time on Monday

**Comment by darkdragon-001** [comment]
It's working fine for me. If I interpret the changes correctly, it just makes sure that the container is deleted before a new one is created. 

If you want to reproduce this in CI, the "correct" way would be to create a dummy docker storage provider which gives gives information like deleted when set with CI. But I guess it's not worth the work... Alternatively emulate very slow storage. But again, I don't know if it's worth to regularly put that much work on the CI. So I would say that manual tests with manual code review should be fine here...

**Comment by WojciechowskiPiotr** [comment]
shipit

**Comment by felixfontein** [comment]
@darkdragon-001 thanks a lot for testing this fix!
@Andersson007 @WojciechowskiPiotr thanks a lot for reviewing!

**Comment by Andersson007** [comment]
@felixfontein , thanks!
ps, people are waiting for the new option added in 2.10:)

**Comment by felixfontein** [comment]
I've created a PR for that option here: #66144

I also found a bug (I think), namely the module shouldn't wait during check mode. I've fixed it here: #66145

**Pull request merged**